src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Sbom.Api.Entities;
+using Microsoft.Sbom.Api.Manifest.FileHashes;
+using Microsoft.Sbom.Contracts.Enums;
+using Microsoft.Sbom.Extensions.Entities;
+using System;
+using System.Linq;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Microsoft.Sbom.Api.Executors
+{
+    /// <summary>
+    /// Validates hashes from multiple file locations (on disk and inside SBOM) simulatenously using
+    /// a conncurrent dictionary.
+    /// </summary>
+    public class ConcurrentHashValidator
+    {
+        private readonly FileHashesDictionary fileHashesDictionary;
+
+        public ConcurrentHashValidator(FileHashesDictionary fileHashesDictionary)        
+        {
+            this.fileHashesDictionary = fileHashesDictionary;
+        }
+
+        public (ChannelReader<FileValidationResult> output, ChannelReader<FileValidationResult> errors)
+            Validate(ChannelReader<InternalSBOMFileInfo> fileWithHash)
+        {
+            var output = Channel.CreateUnbounded<FileValidationResult>();
+            var errors = Channel.CreateUnbounded<FileValidationResult>();
+
+            Task.Run(async () =>
+            {
+                await foreach (var fileHash in fileWithHash.ReadAllAsync())
+                {
+                    await Validate(fileHash, output, errors);
+                }
+
+                output.Writer.Complete();
+                errors.Writer.Complete();
+            });
+
+            return (output, errors);
+        }
+
+        private async Task Validate(InternalSBOMFileInfo internalFileInfo, Channel<FileValidationResult> output, Channel<FileValidationResult> errors)
+        {
+            var sha256Checksum = internalFileInfo.Checksum.Where(c => c.Algorithm == AlgorithmName.SHA256).FirstOrDefault();
+            var fileHashes = new FileHashes();
+            fileHashes.SetHash(internalFileInfo.FileLocation, sha256Checksum);
+            FileValidationResult failureResult = null;
+
+            var newValue = fileHashesDictionary.FileHashes.AddOrUpdate(internalFileInfo.Path, fileHashes, (key, oldValue) =>
+            {
+                // This means a file with the same location was already added to the dictionary.
+                if (oldValue.GetHash(internalFileInfo.FileLocation) != null)
+                {
```

Comment:
When can we find this case?

Reply:
If I have a SBOM that has two entries for the same file path, we will hit this error.